src/components/TodoItem/TodoItem.tsx

```diff
@@ -0,0 +1,162 @@
+import React, { useState, useRef, useEffect } from 'react';
+import classNames from 'classnames';
+import { Todo } from '../../types/Todo';
+import { updateTodos } from '../../api/todos';
+import { ErrorMessages } from '../../types/ErrorMessages';
+import { Loader } from '../Loader/Loader';
+
+type TodoItemProps = {
+  todo: Todo;
+  loading: number[];
+  handleDeleteTodo: (todoId: number) => void;
+  setTodos: (todos: Todo[] | ((prevTodos: Todo[]) => Todo[])) => void;
+  setErrorMessage: (error: string) => void;
+  setLoading: (
+    loading: number[] | ((prevLoading: number[]) => number[]),
+  ) => void;
+};
+
+export const TodoItem: React.FC<TodoItemProps> = ({
+  todo,
+  loading,
+  handleDeleteTodo,
+  setTodos,
+  setErrorMessage,
+  setLoading,
+}) => {
+  const [isChangeState, setIsChangeState] = useState(false);
+  const [updateTodoTitle, setUpdateTodoTitle] = useState(todo.title);
+  const [isSubmitting, setIsSubmitting] = useState(false);
+  const input = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (isChangeState) {
+      input.current?.focus();
+    }
+  }, [isChangeState]);
+
+  const handleToggleTodo = () => {
+    setLoading(prev => [...prev, todo.id]);
+
+    updateTodos(todo.id, { completed: !todo.completed })
+      .then(updatedTodo => {
+        setTodos(prevTodos =>
+          prevTodos.map(todoItem =>
+            todoItem.id === todo.id ? updatedTodo : todoItem,
+          ),
+        );
+      })
+      .catch(() => setErrorMessage(ErrorMessages.ToggleTodo))
+      .finally(() => {
+        setLoading(prev => prev.filter(id => id !== todo.id));
+      });
+  };
+
+  const handleTodoDoubleClick = () => {
+    setIsChangeState(true);
+  };
+
+  const changeTodo = () => {
+    const trimmedTitle = updateTodoTitle.trim();
+
+    if (trimmedTitle === todo.title) {
```

Comment:
Checklist 2. Use destructuring wherever possible. It makes code more readable.
for todo and ErrorMessage)
please check also in another folders